### DISCUSSION TITLE
Build fails  fatal: unable to open include file `reg_sizes.asm'

### DISCUSSION THREAD

**Discussion by Chopin96** [ClickHouse/ClickHouse#50127] [Q&A] [answered]
When I tried to build using the command specified in https://clickhouse.com/docs/en/development/build, I get the following errors. Has anybody also encountered this? 
[0/2] Re-checking globbed directories...
[7440/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc_multibinary.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc_multibinary.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc_multibinary.asm.o ../contrib/isa-l/crc/crc_multibinary.asm
../contrib/isa-l/crc/crc_multibinary.asm:33: fatal: unable to open include file `reg_sizes.asm'
[7441/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_01.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_01.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_01.asm.o ../contrib/isa-l/crc/crc16_t10dif_01.asm
../contrib/isa-l/crc/crc16_t10dif_01.asm:45: fatal: unable to open include file `reg_sizes.asm'
[7442/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc32_gzip_refl_by8.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc32_gzip_refl_by8.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc32_gzip_refl_by8.asm.o ../contrib/isa-l/crc/crc32_gzip_refl_by8.asm
../contrib/isa-l/crc/crc32_gzip_refl_by8.asm:56: fatal: unable to open include file `reg_sizes.asm'
[7443/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_02.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_02.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_02.asm.o ../contrib/isa-l/crc/crc16_t10dif_02.asm
../contrib/isa-l/crc/crc16_t10dif_02.asm:45: fatal: unable to open include file `reg_sizes.asm'
[7445/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_by4.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_by4.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_by4.asm.o ../contrib/isa-l/crc/crc16_t10dif_by4.asm
../contrib/isa-l/crc/crc16_t10dif_by4.asm:46: fatal: unable to open include file `reg_sizes.asm'
[7446/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_copy_by4_02.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_copy_by4_02.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_copy_by4_02.asm.o ../contrib/isa-l/crc/crc16_t10dif_copy_by4_02.asm
../contrib/isa-l/crc/crc16_t10dif_copy_by4_02.asm:47: fatal: unable to open include file `reg_sizes.asm'
[7447/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_copy_by4.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_copy_by4.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_copy_by4.asm.o ../contrib/isa-l/crc/crc16_t10dif_copy_by4.asm
../contrib/isa-l/crc/crc16_t10dif_copy_by4.asm:47: fatal: unable to open include file `reg_sizes.asm'
[7448/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_by16_10.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_by16_10.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc16_t10dif_by16_10.asm.o ../contrib/isa-l/crc/crc16_t10dif_by16_10.asm
../contrib/isa-l/crc/crc16_t10dif_by16_10.asm:48: fatal: unable to open include file `reg_sizes.asm'
[7449/11491] Building ASM_NASM object contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc32_gzip_refl_by8_02.asm.o
FAILED: contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc32_gzip_refl_by8_02.asm.o
/usr/bin/nasm -I../contrib/isa-l/include -I../contrib/isa-l/igzip -I../contrib/isa-l/crc -I../contrib/isa-l/erasure_code -I../base/glibc-compatibility/memcpy -I../contrib/llvm-project/libcxxabi/include -I../contrib/libunwind/include -D OS_LINUX -Werror -f elf64 -o contrib/isa-l-cmake/CMakeFiles/_isal.dir/__/isa-l/crc/crc32_gzip_refl_by8_02.asm.o ../contrib/isa-l/crc/crc32_gzip_refl_by8_02.asm
../contrib/isa-l/crc/crc32_gzip_refl_by8_02.asm:56: fatal: unable to open include file `reg_sizes.asm'
ninja: build stopped: subcommand failed.

**Comment by DanRoscigno** [comment]
@Chopin96 what OS are you running, and which command specifically are you running?

**Reply by Chopin96** [reply]
The commands I used are 
cmake -S . -B build
cmake --build build
I ran this on Ubuntu subsystem for windows version 18.04

**Comment by DanRoscigno** [comment, marked as answer]
Sorry @Chopin96 , the [minimum version of Ubuntu](https://clickhouse.com/docs/en/development/build#building-on-ubuntu) for building ClickHouse is 22.04 

Can you update the WSL image, or do you have Docker installed?

**Reply by Chopin96** [reply]
Thanks for the answer, I will try updating my ubuntu version.

**Reply by DanRoscigno** [reply]
ok, while you are looking into that, I am running a build in Docker on Ubuntu 22.04 (I build on Fedora, but our CI uses Ubuntu and our docs recommend Ubuntu).

**Reply by DanRoscigno** [reply]
I updated https://clickhouse.com/docs/en/development/build with a couple of additional prerequisites, so refresh the page.  My build in Ubuntu 22.04 is still running; I think it took a few hours the first time I did it on my laptop, so it might be a while, but no errors yet.

**Comment by DanRoscigno** [comment]
With an Ubuntu 22.04 Docker container here is my full history:

Note: I am logged in as root

```bash
apt-get update
apt-get install git cmake ccache python3 ninja-build nasm yasm gawk lsb-release wget software-properties-common gnupg sudo

# I did not have sudo installed, if you have sudo use the command in the docs
wget -O - https://apt.llvm.org/llvm.sh | bash

# This is probably only needed if you have multiple versions of clang installed.  If you set it, make sure it is the version that you have
export CXX=clang++-15
export CC=clang-15

# as I do not have my ssh key in the container I have to use the https endpoint for git
 git clone --recursive --shallow-submodules https://github.com/ClickHouse/ClickHouse.git

  cd ClickHouse
  mkdir build
  cmake -S . -B build
  cmake --build build
  
# verify that the programs are built and run
 cd build/
 ls
 ./programs/clickhouse-client 
```

**Reply by Chopin96** [reply]
Hi Dan, the build was a success after updating my ubuntu to version 22.04. Thanks for helping!

**Comment by DanRoscigno** [comment]
Excellent!

**Comment by shiyu15** [comment]
i solved the problem by upgrading the nasm version. Ubuntu 18.04 has nasm-2.13 as its default nasm version. with this version, these errors may happen. with Ubuntu 20.04's nasm-2.14.02, there are no errors. So we can install nasm-2.14.02 manually.   
1. download from https://www.nasm.us/pub/nasm/releasebuilds/2.14.02/
2. compile and install
```bash
tar xf nasm-2.14.02.tar.bz2

cd nasm-2.14.02/

BUILD_LIBS=${HOME}/build_libs

./configure --prefix=${BUILD_LIBS}

make

make install
```
3. cmake the clickhouse repo again, if there are errors like 
```bash
CMake Error at contrib/isa-l-cmake/CMakeLists.txt:28 (enable_language):
  The CMAKE_ASM_NASM_COMPILER:

    /usr/bin/nasm

  is not a full path to an existing compiler tool.
```

in nasm directory, install nasm to the /usr/bin/nasm
```bash
sudo /usr/bin/install -c nasm /usr/bin/nasm
sudo /usr/bin/install -c ndisasm /usr/bin/ndisasm
```
